File: Problemas/cpp/functors/java/MainPredicate.java
import java.util.Collection;
import java.util.List;
import java.util.LinkedList;
import java.util.ArrayList;

class MainPredicate
{
	public static void main(String[] argv)
	{
		Functors.select(new Functors.IsEven(),1,10);
		Functors.select(new Functors.IsPrime(),1,30);

		try
		{
			List<Integer> l = new LinkedList(List.of(1,2,3,4,5,6,7,8,9,10));
			out(l);

			var lfe = Functors.filter(new Functors.IsEven(),l);
			out(lfe);
			var lfp = Functors.filter(new Functors.IsPrime(),l);
			out(lfp);

			var lme = Functors.map(new Functors.IsEven(),l);
			out(lme);
			var lmp = Functors.map(new Functors.IsPrime(),l);
			out(lmp);

			List<Integer> v = new ArrayList(List.of(1,2,3,4,5,6,7,8,9,10));
			out(v);

			var vfe = Functors.filter(new Functors.IsEven(),v);
			out(vfe);
			var vfp = Functors.filter(new Functors.IsPrime(),v);
			out(vfp);

			var vme = Functors.map(new Functors.IsEven(),v);
			out(vme);
			var vmp = Functors.map(new Functors.IsPrime(),v);
			out(vmp);
		}
		catch(Exception e)
		{
			e.printStackTrace();
		}
	}

	static <T> void out(Collection<T> c)
	{
		System.out.println(c.getClass().getName()+": "+c);
	}
}

File: Problemas/cpp/functors/c++/main-predicate.cc
#include <iostream>
#include "functor.h"
#include <list>
#include <vector>

using namespace std;
template<typename T>
void out(const T& c)
{
	for (const auto& e : c)
		cout << boolalpha << e << " ";
	cout << endl;
}

int main()
{
	fn::select(fn::is_even(),1,10);
	fn::select(fn::is_prime(),1,30);

	list<int>  l {1,2,3,4,5,6,7,8,9,10};

	auto lfe = fn::filter(fn::is_even(),l);
	out(lfe);
	auto lfp = fn::filter(fn::is_prime(),l);
	out(lfp);

	auto lme = fn::map(fn::is_even(),l);
	out(lme);
	auto lmp = fn::map(fn::is_prime(),l);
	out(lmp);

	vector<int>  v {1,2,3,4,5,6,7,8,9,10};

	auto vfe = fn::filter(fn::is_even(),v);
	out(vfe);
	auto vfp = fn::filter(fn::is_prime(),v);
	out(vfp);

	auto vme = fn::map(fn::is_even(),v);
	out(vme);
	auto vmp = fn::map(fn::is_prime(),v);
	out(vmp);

	return 0;
};

